Added Host Sublets page

nathanclairmonte opened the pull request:
closes #17 

Did the following things in this PR:

## Host Sublets Page
- Added the `/pages/host/sublets.js` page, which is visually identical to the regular `/pages/sublets.js` page, except that now it checks for listings the user has created themselves (instead of checking for listings that the user has created requests for).
- The logic here is identical to the one on the CRA app. If we ever want to change in the future we can, but I decided to keep the old logic for now until we discuss that.
  - First we fetch all the requests in the DB as well as all the published and expired listings for the currently signed in user (if user isn't signed in they'll just get a guest page).
  - For all the published and expired listings we fetch, we filter through all the requests and attach the ones for that listing to the listing object in a `.requests` attribute.
  - Then, the displayed listings are as follows:
    - In the "active" tab, all the published listings with `request.status !== "confirmed"` are shown
    - In the "past" tab, all the expired listings are shown
    - In the "confirmed" tab, all the published listings with `request.status === "confirmed"` are shown
- Finally, the only way to get to this page is through the "Switch to Hosting" button on the profile page. Added this link as well.

## Host Listing Page
- Added the `/pages/host/listing/[listingId].js` page, which is also visually very similar to the regular `/pages/listing/[listingId].js` page, with a few small differences.
- Kept the image carousel at the top, as well as the title, active bid info, address, user and room information. Changed the price to just shown the current listing price.
- The other stuff is removed (stuff like amenities displays, location, legal stuff etc.)
- Then, added a request list which just shows a list of all requests for that listing (took this concept from CRA app as well)

## API Routes
- Added `/api/requests/index.js` to get all requests in the DB. Using this on the new host sublets page
- Added `/api/listings/published/[userId].js` to get all published listings for a given user
- Added `/api/listings/expires/[userId].js` to get all expired listings for a given user

## Components
- `ListingItemForHostSublets.js`
  - Another `ListingItem`-esque component, with the main difference being that this one now links to the host listing page (instead of the regular listing page)
  - Also added a "views" component to show how many views the listing has.

## Miscellaneous
- Updated the `BottomNav` component to highlight the "Sublets" icon if we're on the host sublets page as well as the regular sublets page

cemgunay reviewed (commented):
Yo this is hellllllla clean i really like how the rendering works with the different requests its very simple and nice and attaching requests to a listing. Here are my thoughts on some changes let me know what you think!

### Host Listing Page
- Add an edit Icon (like a pencil or something) in the top right if its an active/past listing that takes user to `host/manage/listings/${listingId}/edit (doesnt exist in this branch but i created it so when we merge it should be gucci)
- Have 2 seperate tabs for active and past requests both sorted by updatedAt descending by default
  - Eventually we will add option to sort by request price also if you want to do that now but not a priority imo
- Add first name of requestee in line with the price justified between
- When user clicks a request it goes to that request page but for host
  - This will need to be created, very similar to the subtenant request page but just shows the request. We can add this when we are doing the request process, just putting it as a note here for later

### Host Sublets Page
Currently listings are being categorized as follows:
- Active listings: Published listings without confirmed requests.
- Confirmed listings: Published listings with confirmed requests.
- Past listings: Listings that have expired.

However we can simplify this since the workflows we will create (for booking a listing and whatnot) will adjust the listing records' properties accordingly. 
So for example when a request is confirmed the attached listings isBooked property becomes true and published property becomes false. I forgot to write about this in the document so my badddddd. 

So we can simplify to:
- Active listings: Listings with published true
- Confirmed listings: Listing with isBooked true
- Past listings: Listings that have expired.

I really like the idea of attaching requests to listings like you did. This is just a suggestion I thought of when looking at that for optimization but instead of calling all requests and then filtering and attaching in the client we can do an aggregated join using $lookup to join relevant requests to each listing which can be more efficient to handle it all in the server side.

For example for published listings

```
import connectMongo from "@/utils/connectMongo";
import mongoose from "mongoose";
import Listing from "@/models/Listing";
import Request from "@/models/Request";

// get all published listings for a user with their requests
export default async function handler(req, res) {
    // switch based on request method
    switch (req.method) {
        case "GET":
            // get user ID from query
            const { userId } = req.query;

            // check if the userId is a valid Mongo ObjectId
            if (!mongoose.Types.ObjectId.isValid(userId)) {
                return res.status(404).json({ error: "User ID not valid" });
            }

            try {
                // connect to DB
                await connectMongo();

                // get all published listings for the user with their requests
                const listingsWithRequests = await Listing.aggregate([
                    { $match: { userId: mongoose.Types.ObjectId(userId), published: true } },
                    { $lookup: {
                        from: 'requests', // the collection to join
                        localField: '_id', // field from the listings collection
                        foreignField: 'listingId', // field from the requests collection
                        as: 'requests' // the array field to hold the join result
                    }}
                ]);

                // return the listings with requests with 200 status code
                res.status(200).json(listingsWithRequests);
            } catch (error) {
                console.error(`Failed to retrieve listings with requests:\n`, error);
                res.status(500).json({ error: "Internal Server Error" });
            }
        default:
            // got some method other than GET
            res.setHeader("Allow", ["GET"]);
            res.status(405).end(`Method ${req.method} Not Allowed`);
    }
}

```

### ListingItemForHostSublets
For the "active" tab:
- If all requests are rejected it displays a message indicating "No Offers."
- If there is a highest active request price (highestActiveRequestPrice), it displays the highest offer
- If neither condition is met, it shows a default error message ("SOMETHING'S WRONG").
However the scenario where no offers have been placed yet (so nothing has been rejected) is not covered it will still show somethings wrong when it should say no offers

And I saw your comment about handling what it should say for past and confirmed heres what I think:

For the "past" tab:
- Show nothing lol lets keep it simple idk

For the "confirmed" tab:
- The amount of the request that was confirmed

Also do we need to still do an api call here to get the requests? With the new logic aren't they already coming from the requests prop?

### BottomNav
For future, should we have a seperate BottomNavHost that takes us to host links so we can navigate within host?

### Misc
- I am going to add an index on listingId and userId in the requests table since we never really just want all requests, it can be inefficient and eventhough putting indexes like this might make the write and put operations heavier, I think it will still be beneficial let me know what you think.
- Along the same thought process adding an index on userId in listings table might also be beneficial

nathanclairmonte commented:
These are great suggestions!! Love the idea of attaching the requests on the server side that's brilliant

Here are the changes I made from these suggestions:
- Added an edit button to the top right of the `/host/listing/[listingId].js` page that links to `host/manage/listings/${listingId}/edit` if it's an active/past listing.
- Added a `RequestItemForHostListing` component, so that each individual request can have a useEffect to fetch the user info of the subTenant of that request (so we can get first name to display)
  - Just to confirm, when you say requestee you mean the subTenant right? not the tenant? The former made more sense to me, like we want the name of the person who submitted the request.
  - If not though, it's an easy change. We'd just call the user endpoint with `request.tenandId` instead of `request.subTenantId` like we are doing now.
- Added a general `Tabs` component, where you can give it a list of tab options (up to 6) and it will create the tabs dynamically (instead of hardcoding the tab options like we did with SubletsTabs). If you like this idea I can update the sublets pages (host and regular) to use this general tabs component instead of the `SubletsTabs` component too.
- Now the requests on each host listing page will be divided into active and past requests, based on the `ACTIVE_STATUSES` and `PAST_STATUSES` arrays we have previously defined.
- Added a new API route to get listings for a user, exactly the same as the published and expired ones, but this time for booked listings.
- Updated the active/past/confirmed logic on the host sublets page as you suggested:
  - Active listings are ones where `published = true` (using the `/api/listings/published/[userId]` API route)
  - Past listings are ones that are expired (using the `/api/listings/expired/[userId]` API route)
  - Confirmed listings are ones where `isBooked = true` (using the `/api/listings/booked/[userId]` API route)
  - No longer fetching all requests
- All the API routes mentioned above now attach the requests to each listing on the server side. Great suggestion I love this!
- On the `ListingItemForHostSublets` component, I changed the price text to show "No Offers" for the scenario where no offers have been placed yet. Also shows nothing for past tab and the amount of the request that was confirmed for confirmed tab
- Took out the API call to get the requests bc ur right they're coming in with the requests prop! Just filtered for the active ones instead of calling the `/listingactiverequests` API route (on both the `ListingItemForHostSublets` and the host sublets page). This is defff better. One thing to note tho is that one of the listings I was looking at in testing went from "1 active bid" to "no active bids" when I used the passed-in request prop instead of the one from the API route. Not entirely sure why this is happening but I feel like once we streamline all the request logic we'll catch all the bugs like this


Some other thoughts from ur comments:
- I think a separate BottomNavHost is prob the way we're gonna have to go fr! Maybe some sort of `isHostMode` flag that we add to context and/or the user auth object (will be much easier to do with NextAuth too actually) that will determine whether they're in hosting mode or not, and then we display like a whole different bottom nav based on that. Maybe even like different bottom nav labels altogether too to distinguish between host mode and non-host mode?
- Think adding an index for the requests and listings tables is a good idea! You're right we never really want all requests, or even all listings tbh. Right now we're grabbing all listings to displkay on the Explore page but eventually that's gonna include like pagination and various filters for location/dates/prices etc. so it's unlikely we'll be grabbing all listings often.

cemgunay commented:
Sorry I took forever to look at your updates homie, I am finally able to use a keyboard again lmao.

The changes are great, and yeah sorry when i said requestee I meant subTenant so you assumed correct lmao.

For it going from 1 active request to no active bids its because the listingId in Request model is a string and therefore when we are doing the aggregate in the api it cant lookup a string to an ObjectId so i changed the listingId in Requests model to be ObjectId instead of string. Hopefully this doesnt mess up anything when a new request is created but we will see when we start working on that.

For the confirmed tabs prices, you're right it should never reach the else statement since every confirmed booking should have a price. This might be because of some old testing stuff where I bugged around with things even after listings were booked but we will sort all this out like you said after we sort the request/book logic so we can just leave it for now.
